pkg/server/commands/write.go

```diff
@@ -53,6 +53,10 @@ func (c *WriteCommand) validateWriteRequest(ctx context.Context, req *openfgapb.
 	defer span.End()
 
 	store := req.GetStoreId()
+	// Check if store is empty, can't write to empty store
+	if store == "" {
```

Comment:
Instead of checking whether the store is null - as we have already validate it as part of [protobuf API](https://github.com/openfga/api/blob/157c45499d9ba7218b98601a0608281ddc82ee1b/openfga/v1/openfga_service.proto#L809), we should call [GetStore](https://github.com/openfga/openfga/blob/c108f3c5142d3aa8a10909b9ee253d4a4151ea92/pkg/storage/storage.go#L175) to see whether store is returned or the store is non-existing.

Reply:
We'll also want to ensure that we do the GetStore read in the same transaction as the other statements. Otherwise we'll have a race on the store's visibility relative to other concurrent transactions. As part of this work, let's consider adding a `WrapTransaction` or similarly named signature to the storage interface. For example:


```go

type TxnFuncHandler func(ctx context.Context) error

type OpenFGADatstore interface {
    ...
    WrapTransaction(ctx context.Context, txnFunc TxnFuncHandler) error
}
```

then, in calling code, we can check for the store's existence within the same txn

```go

func (cmd *WriteCommand) Write(ctx context.Context, ...) (...) {
    ... // other code
    
    err := cmd.datastore.WrapTransaction(ctx, func(ctx contextContext) error {
        _, err := cmd.datastore.GetStore(ctx, req.GetStoreId())
        if err != nil {
            if errors.Is(err, storage.ErrNotFound) {
                // handle non-existence case
            }

            // handle general error
        }

    ... // handle other Write database code.
    })
}
```

How we compose this is up for debate, but the importance is that we avoid races on the store's visibility relative to the transactions that have committed.

Reply:
> Otherwise we'll have a race on the store's visibility relative to other concurrent transactions

@jon-whit I don't follow, can you give an example of what problems may happen if we don't do a transaction?

Reply:
@miparnisari Sure. Consider this scenario:

Two clients are concurrently interacting with the OpenFGA API. Client 1 calls the `Write` API against `store1`. Meanwhile, client 2 calls the `DeleteStore` API against `store1`. Consider the following code in the `Write` API:

```go
func (cmd *WriteCommand) Execute(ctx context.Context, req *openfgapb.WriteRequest) (*openfgapb.WriteResponse, error) {

    store, err := cmd.datastore.GetStore(ctx, req.GetStoreId()) // step 1
    if err != nil {
        if errors.Is(err, storage.ErrNotFound) {
            return nil, status.Errorf(codes.InvalidArgument, "the provided store_id does not exist")
        }
    
        return nil, err
    }

    // other code for validation etc..                         // step 2

   err = cmd.datastore.Write(....)                             // step 3
   if err != nil {
       // handle error
   }
}
```

If client 2 concurrently deletes the store between step 1 and step 2, then the Write at step 3 is erroneous. The set of rows satisfying the prior query condition (GetStore) may have changed due to another recently-committed transaction. This is known as a phantom read (from the perspective of the API client). For data consistency and API consistency reasons, API behavior should behave in a serializable manner in regards to the client. If a client is concurrently deleting a store while another client is concurrently writing to it, then whichever transaction commits first should win. Otherwise you get concurrency anomalies. 

See https://www.postgresql.org/docs/current/transaction-iso.html for more information, but we should be using an isolation level of at-least "repeatable-read" (the default in postgres), but that only applies to a transaction itself. If you call `datastore.GetStore` followed by `datastore.Write`, then those two queries aren't executed in the same transaction, and so a race on the previously read data can happen between step 1 and step 2. It's pretty common to see code written like this in projects, and it's always bothersome because it's not correct and/or transactionally accurate.

Reply:
I kind of agree on above @jon-whit . Do we have examples of similar `Transaction` in repo, otherwise I can add some module which can be reused. 
Cc @miparnisari

Reply:
@Himanshu372, we don't at the moment. The current storage interface (and the way it is consumed) doesn't demand transaction support as part of the storage interface. This would be one of the first use cases of having such a signature added to the storage interface. Like I said above, we should just consider adding a `WrapTransaction` or something similar, and then we can implement that signature to hook into a txn across multiple datastore calls.

Reply:
> If client 2 concurrently deletes the store between step 1 and step 2, then the Write at step 3 is erroneous. 

I feel like we are adding a big change to our storage interface to support one scenario that happens once (the deletion)

Reply:
> I feel like we are proposing a big change to our storage interface to support just one scenario that happens once (the store deletion)

We wouldn't write code with data race conditions on shared memory (e.g. `go test -race ...`), so why would we compromise on design accuracy and consistency on data at the storage layer?

Reply:
@jon-whit @miparnisari  When doing store check for every API which has `store_id` as param, I do see a lot of diff. Adding `Transaction` logic here would be inflate the diff more and PR would be difficult of review. 
I can open up another issue to take that work forward, is it OK?

Reply:
@adriantam as you suggested I have used `GetStore` logic to validate store, plz let me know if it looks good to you